src/main/java/uk/gov/hmcts/darts/event/service/impl/EventHandlerBase.java

```diff
@@ -66,4 +95,107 @@ protected String buildKey(String type, String subType) {
         requireNonNull(type);
         return type + (isNull(subType) ? "" : subType);
     }
+
+    protected CourtroomCourthouseCourtcase getOrCreateCourtroomCourtHouseAndCases(DartsEvent dartsEvent) {
+        final var actualCourtRoomEntity = getCourtRoom(dartsEvent);
+
+        final var caseNumbers = dartsEvent.getCaseNumbers();
+        if (caseNumbers.size() > 1) {
+            log.warn(format(MULTIPLE_CASE_NUMBERS, dartsEvent.getEventId(), join(", ", caseNumbers)));
+        }
+        final var actualCourtHouse = actualCourtRoomEntity.getCourthouse();
+        final var actualEventCaseNumber = caseNumbers.get(0);
+        final var actualEventDate = dartsEvent.getDateTime();
+
+        var actualCourtCaseEntity = getCourtCaseEntityOrCreate(actualCourtHouse, actualEventCaseNumber);
+
+        boolean hearingIsNew = false;
+        HearingEntity actualHearingEntity;
+        var optionalHearingEntity = getHearingEntity(actualEventDate, actualCourtCaseEntity);
+        if (optionalHearingEntity.isPresent()) {
+            actualHearingEntity = optionalHearingEntity.get();
+        } else {
+            actualHearingEntity = newCaseHearing(actualCourtRoomEntity, actualEventDate, actualCourtCaseEntity);
+            hearingIsNew = true;
+        }
+        actualHearingEntity.setHearingIsActual(true);
+
+        saveEvent(dartsEvent, actualCourtRoomEntity);
+
+        boolean isCourtroomDifferentFromHearing = isCourtroomDifferentFromHearingCourtroom(actualCourtRoomEntity, actualHearingEntity);
+
+        if (isTheHearingNewOrTheCourtroomIsDifferent(hearingIsNew, isCourtroomDifferentFromHearing)) {
+            actualHearingEntity.setCourtroom(actualCourtRoomEntity);
+        }
+
+        caseRepository.save(actualCourtCaseEntity);
+        hearingRepository.save(actualHearingEntity);
+
+        return CourtroomCourthouseCourtcase.builder()
+            .courthouseEntity(actualCourtHouse)
+            .courtroomEntity(actualCourtRoomEntity)
+            .caseEntity(actualCourtCaseEntity)
+            .isHearingNew(hearingIsNew)
+            .isCourtroomDifferentFromHearing(isCourtroomDifferentFromHearing)
+            .build();
+    }
+
+    protected CourtroomEntity getCourtRoom(DartsEvent dartsEvent) {
```

Comment:
i think we should be creating a courtroom if its not been found instead of raising an exception.
There is already existing code in CommonApiImpl.retrieveOrCreateCourtroom that could be used, and raises more specific errors, although they are in the old format, so maybe could be updated.

Reply:
This behaviour was agreed upon as part of DMP-153.  Jack wasn't sure we should be following the same logic as CommonApiImpl.retrieveOrCreateCourtroom.  Please catch up with Jack before changing this.  It isn't behaviour introduced as part of this ticket either.